Library/Homebrew/formula_installer.rb

```diff
@@ -1129,24 +1130,60 @@ def forbidden_license_check
                                             .to_s
                                             .sub("Public Domain", "public_domain")
                                             .split(" ")
+                                            .to_h do |license|
+      [license, SPDX.license_version_info(license)]
+    end
+
     return if forbidden_licenses.blank?
 
     compute_dependencies.each do |dep, _|
       next if @ignore_deps
 
       dep_f = dep.to_formula
-      next unless dep_f.license.all? { |license| forbidden_licenses.include?(license.to_s) }
+      next unless licenses_forbid_installation? dep_f.license, forbidden_licenses
 
       raise CannotInstallFormulaError, <<~EOS
-        The installation of #{formula.name} has a dependency on #{dep.name} where all its licenses are forbidden: #{dep_f.license}.
+        The installation of #{formula.name} has a dependency on #{dep.name} where all its licenses are forbidden: #{SPDX.license_expression_to_string dep_f.license}.
       EOS
     end
     return if @only_deps
 
-    return unless formula.license.all? { |license| forbidden_licenses.include?(license.to_s) }
+    return unless licenses_forbid_installation? formula.license, forbidden_licenses
 
     raise CannotInstallFormulaError, <<~EOS
-      #{formula.name}'s licenses are all forbidden: #{formula.license}.
+      #{formula.name}'s licenses are all forbidden: #{SPDX.license_expression_to_string formula.license}.
     EOS
   end
+
+  def licenses_forbid_installation?(license_expression, forbidden_licenses)
+    case license_expression
+    when String, Symbol
+      forbidden_licenses_include? license_expression.to_s, forbidden_licenses
+    when Hash
+      key = license_expression.keys.first
+      case key
+      when :any
+        license_expression[key].all? { |license| licenses_forbid_installation? license, forbidden_licenses }
+      when :all
+        license_expression[key].any? { |license| licenses_forbid_installation? license, forbidden_licenses }
+      else
+        forbidden_licenses_include? key, forbidden_licenses
+      end
+    end
+  end
+
+  def forbidden_licenses_include?(license, forbidden_licenses)
+    return true if forbidden_licenses.key? license
+
+    name, version, = SPDX.license_version_info license
+
+    forbidden_licenses.each do |_, license_info|
+      next unless license_info[0] == name
+
+      return true if license_info[2] && license_info[1] <= version
+
+      return true if license_info[1] == version
+    end
+    false
+  end
```

Comment:
I'd like to have some tests for this logic. Is there an easy way to test these methods in [`formula_installer_spec.rb`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/formula_installer_bottle_spec.rb)?

Otherwise, I can extract them out to the [`SPDX` module](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/utils/spdx.rb).

Reply:
Yeh, I'd suggest moving them all to `SPDX`.